miniapps/shifted_methods/distance.cpp

```diff
@@ -0,0 +1,285 @@
+//
+//   Problem 3: level set - Gyroid.
+//      mpirun -np 4 distance -m ../../data/periodic-square.mesh -rs 5 -o 2 -t 1.0 -p 3
+//      mpirun -np 4 distance -m ../../data/periodic-cube.mesh -rs 3 -o 2 -t 1.0 -p 3
+//
+
+#include <fstream>
+#include <iostream>
+#include "dist_solver.hpp"
+
+using namespace std;
+using namespace mfem;
+
+double sine_ls(const Vector &x)
+{
+   const double sine = 0.25 * std::sin(4 * M_PI * x(0)) +
+                       0.05 * std::sin(16 * M_PI * x(0));
+   return (x(1) >= sine + 0.5) ? -1.0 : 1.0;
+}
+
+double sphere_ls(const Vector &x)
+{
+   const int dim = x.Size();
+   if (dim == 2)
+   {
+      const double xc = x(0) - 0.5, yc = x(1) - 0.5;
+      const double r = sqrt(xc*xc + yc*yc);
+      return (r >= 0.4) ? -1.0 : 1.0;
+   }
+   else if (dim == 3)
+   {
+      const double xc = x(0) - 0.0, yc = x(1) - 0.0, zc = x(2) - 0.0;
+      const double r = sqrt(xc*xc + yc*yc + zc*zc);
+      return (r >= 0.8) ? -1.0 : 1.0;
+   }
+   else
+   {
+      return (x(0) >= 0.5) ? -1.0 : 1.0;
+   }
+}
+
+double Gyroid(const Vector &xx)
+{
+   const double period = 2.0 * M_PI;
+   double x=xx[0]*period;
+   double y=xx[1]*period;
+   double z=0.0;
+   if(xx.Size()==3)
+   {
+      z=xx[2]*period;
+   }
+   return std::sin(x)*std::cos(y) +
+          std::sin(y)*std::cos(z) +
+          std::sin(z)*std::cos(x);
+}
+
+double Sph(const mfem::Vector &xx)
+{
+    double R=0.4;
+    mfem::Vector lvec(3);
+    lvec=0.0;
+    for(int i=0;i<xx.Size();i++)
+    {
+        lvec[i]=xx[i];
+    }
+
+    return lvec[0]*lvec[0]+lvec[1]*lvec[1]+lvec[2]*lvec[2]-R*R;
+}
+
+void DGyroid(const mfem::Vector &xx, mfem::Vector &vals)
+{
+    vals.SetSize(xx.Size());
+    vals=0.0;
+
+    double pp=4*M_PI;
+
+    mfem::Vector lvec(3);
+    lvec=0.0;
+    for(int i=0;i<xx.Size();i++)
+    {
+        lvec[i]=xx[i]*pp;
+    }
+
+    vals[0]=cos(lvec[0])*cos(lvec[1])-sin(lvec[2])*sin(lvec[0]);
+    vals[1]=-sin(lvec[0])*sin(lvec[1])+cos(lvec[1])*cos(lvec[2]);
+    if(xx.Size()>2)
+    {
+        vals[2]=-sin(lvec[1])*sin(lvec[2])+cos(lvec[2])*cos(lvec[0]);
+    }
+
+    vals*=pp;
+}
+
+int main(int argc, char *argv[])
+{
+   // 1. Initialize MPI.
+   int num_procs, myid;
+   MPI_Init(&argc, &argv);
+   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+   MPI_Comm_rank(MPI_COMM_WORLD, &myid);
+
+   // 2. Parse command-line options.
+   const char *mesh_file = "../data/star.mesh";
+   int solver_type = 0;
+   int problem = 0;
+   int rs_levels = 0;
+   int order = 2;
+   double t_param = 1.0;
+   const char *device_config = "cpu";
+   bool visualization = true;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&solver_type, "-s", "--solver",
+                  "Solver type:\n\t"
+                  "0: Heat\n\t"
+                  "1: P-Laplacian");
+   args.AddOption(&problem, "-p", "--problem",
```

Comment:
It looks like this description is outdated since the sample runs use values from 0 to 3.

Reply:
Fixed.